concent_api/core/payments/bankster.py

```diff
@@ -103,6 +103,13 @@ def claim_deposit(
     # Bankster puts database locks on DepositAccount objects
     # that will be used as payers in newly created DepositClaims.
     with non_nesting_atomic(using='control'):
+        DepositAccount.objects.select_for_update().filter(
+            client=requestor_client,
+            ethereum_address=requestor_ethereum_address,
+        ).filter(
+            client=provider_client if is_claim_against_provider else None,
```

Comment:
`client` is defined as `client = ForeignKey(Client)` so it can't be `None`. This way you won't ever match any object. You have to add this filter condition conditionally.

Reply:
It is not a matter of clean code and it will not work correctly. `filter` in Django works as SQL `AND` operator, so if you have ie.
```
In [6]: Client.objects.filter(public_key='ZlXidtPmabypX/uTKl8e7drZcGrwyrEcfsHvxMNc6RAiWWfpJX9/FgPLdEW3HueIqK6gUKgyYuQ9KSnvL4McUg==')
Out[6]: <QuerySet [<Client: Client object>]>
```

This returns object, but when you add 

```
In [7]: Client.objects.filter(public_key='ZlXidtPmabypX/uTKl8e7drZcGrwyrEcfsHvxMNc6RAiWWfpJX9/FgPLdEW3HueIqK6gUKgyYuQ9KSnvL4McUg==').filter(public_key=None)
Out[7]: <QuerySet []>
```

This returns empty result, so even the first matching result will not be locked. So you have split it into two queries, or add second filter conditionally.

Reply:
I agree with @rwrzesien. It works like an AND. It's possible to achieve OR with `Q()` expressions but I think that doing two separate queries makes it more apparent that two separate objects may be locked.

Locking them in a single operation does have one advantage though - it makes it harder to cause a deadlock. When two processes (X and Y) need to lock two separate objects (A and B) there's a risk that X will lock A, Y will lock B and then X will be waiting for B while Y will be waiting for A - indefinitely. That's a deadlock. To avoid this you have to always lock objects in the same order - i.e. both X and Y should try to lock A first and only when that succeeds try to lock B. Locking both in a single operation is another solution.

Please look at all the other functions that may lock two objects and make sure that they either try to acquire locks in the same order or lock both in a single operation.

Reply:
OK, i will do it in two separate queries. My doubt is only that we have two database queries this way and it could be achieved with only one (OR with `Q()` as you wrote). I do it as you wrote, 2 queries and if you would decide to make it in one please let me know.

Reply:
Should be OK with `OR` and will be good exercise to get familiar how to construct such query.

Reply:
One more query per HTTP request is insignificant. Now worth worrying about. You should start worrying when the number of queries depends on the size of the input or data stored in the database. I.e. when you have a list and have to make a separate query for each item.

If you want to learn more about database performance problems in Django, here are some articles:
- [All you need to know about prefetching in Django](https://hackernoon.com/all-you-need-to-know-about-prefetching-in-django-f9068ebe1e60)
- [Solving Performance Problems in the Django ORM](https://medium.com/@hansonkd/performance-problems-in-the-django-orm-1f62b3d04785)

In this case either way of doing it - one query or two - is fine for me. Each one has its pros and cons. The important thing was the bug pointed out by @rwrzesien (doing an AND operation). Now that it's fixed, the rest is just a matter of preference and readability.

Reply:
Thanks, I will read this. I like this part of programming :)